### PR TITLE
Add test commands to search and survey rule hits

### LINKED ISSUE #484: [FR] Add commands to search with rule queries
Add commands to assist with searching for results against rule queries against a stack, to assist with testing.

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
resolves #484

## Summary
Adds the following commands:

* `dev test`:
   - `event-search` - search elasticsearch with a specified query
   - `rule-event-search` - search elasticsearch by specifying a rule
   - `rule-survey` - run a survey against all rules specified by a `rule-search` filter and determine search and alert hits
* `kibana`
   - `search-alerts` - search alerts with kql

```console
python -m detection_rules dev test rule-survey --hide-zero-counts --hide-errors --date-range now-220d now
Loaded config file: .../.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

DEBUG MODE ENABLED
Running survey against 325 rules
Saving detailed dump to: surveys/20201106T014126L.json
===============================================================================================================================
 rule_id                                name                                                        search_count   alert_count 
===============================================================================================================================
 2bf78aa2-9c56-48de-b139-f169bf99cf86   Adobe Hijack Persistence                                               1    
 6ea71ff0-9e95-475b-9506-2580d1ce6154   DNS Activity to the Internet                                         552    
 f675872f-6d85-40a3-b502-c0d2ef101e92   Delete Volume USN Journal with Fsutil                                  1    
 4b438734-3793-4fda-bd42-ceeada0be8f9   Disable Windows Firewall Rules via Netsh                               2    
 9c260313-c811-4ec8-ab89-8f6530e0246c   Hosts File Modified                                                   31    
 afcce5ad-65de-4ed2-8516-5e093d3ac99a   Local Scheduled Task Commands                                          8    
 e8571d5f-bea1-46c2-9f56-998de2d3ed95   Local Service Commands                                               107   12 
 adb961e0-cb74-42a0-af9e-29fc41f88f5f   Netcat Network Activity                                                1    
 81cc58f5-8062-49a2-ba84-5cc4b4d31c40   Persistence via Kernel Module Modification                             1    
 68921d85-d0dc-48b3-865f-43291ca2c4f2   Persistence via TelemetryController Scheduled Task Hijack             23    
 7405ddf1-6c8e-41ce-818f-48bea6bcaed8   Potential Modification of Accessibility Binaries                      25    
 6ea41894-66c3-4df7-ad6b-2c5074eb3df8   Process Potentially Masquerading as WerFault                           1    
 b41a13c6-ba45-4bab-a534-df53d0cfed6a   Suspicious Endpoint Security Parent Process                            7    
 a624863f-a70d-417f-a7d2-7a404638d47f   Suspicious MS Office Child Process                                     0   12 
 fd7a6052-58fa-4397-93c3-4795249ccfa2   Svchost spawning Cmd                                                   2    
 c7ce36c0-32ff-4f9a-bfc2-dcb242bf99f9   Unusual File Modification by dns.exe                                 130    
 3b47900d-e793-49e8-968f-c90dc3526aa1   Unusual Parent Process for cmd.exe                                    64    
 35df0dd8-092d-4a83-88c1-5151a804f31b   Unusual Parent-Child Relationship                                      3    
 df959768-b0c9-4d45-988c-5606a2be8e5a   Unusual Process Execution - Temp                                     104    
===============================================================================================================================
```

### CODE REVIEW

File: detection_rules/devtools.py [hunk 1]
```diff
@@ -201,3 +206,145 @@ def license_check(ctx):
...
+
+
+@dev_group.group('test')
+def test_group():
+    """Commands for testing against stack resources."""
+
+
+@test_group.command('event-search')
+@click.argument('query')
+@click.option('--index', '-i', multiple=True, required=True, help='Index(es) to search against ("*": for all indexes)')
```
Comment (rw-access) [inline detection_rules/devtools.py:218]:
```suggestion
@click.option('--index', '-i', multiple=True, required=False, help='Index patterns to search against')
```

File: detection_rules/devtools.py [hunk 2]
```diff
@@ -201,3 +206,145 @@ def license_check(ctx):
...
+@click.option('--max-results', '-m', type=click.IntRange(1, 1000), default=100,
+              help='Max results to return (capped at 1000)')
+@click.option('--verbose', '-v', is_flag=True, default=True)
+@add_client('elasticsearch')
+def event_search(query, index, language, date_range, count, max_results, verbose=True,
+                 elasticsearch_client: Elasticsearch = None):
+    """Search using a query against an Elasticsearch instance."""
+    start_time, end_time = date_range
+    language_used = "kql" if language is None else "eql" if language is True else "lucene"
+    collector = CollectEvents(elasticsearch_client, max_results)
```
Comment (rw-access) [inline detection_rules/devtools.py:231]:
```suggestion
    collector = CollectEvents(elasticsearch_client, max_results)
    index = index or ("*", )
```

File: detection_rules/devtools.py [hunk 3]
```diff
@@ -201,3 +206,145 @@ def license_check(ctx):
...
+        click.echo(f'total results: {len(results)} (capped at {max_results})')
+        click.echo_via_pager(json.dumps(results, indent=2, sort_keys=True))
+
+    return results
+
+
+@test_group.command('rule-event-search')
+@click.argument('rule-file', type=click.Path(dir_okay=False), required=False)
+@click.option('--rule-id', '-id')
+@click.option('--date-range', '-d', type=(str, str), default=('now-7d', 'now'), help='Date range to scope search')
```
Comment (rw-access) [inline detection_rules/devtools.py:251]: wait what. this is a thing?

**Comment by threat-punter** [comment]
I tested the `event-search` and `rule-survey` commands successfully. Very nice.

I saw the error below when testing the `rule-event-search` command though.

```
(.venv) hostname:detection-rules username$ python -m detection_rules dev test rule-event-search --cloud-id "my-cloud-id" --rule-id "e48236ca-b67a-4b4e-840c-fdc7782bc0c3"
█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█
es_user: elastic
es_password:
CLI Error (NoneType): Missing required --cloud-id or --elasticsearch-url
```

**Review by threat-punter — CHANGES_REQUESTED** [review]
I tested the `event-search` and `rule-survey` commands successfully. Very nice.

I saw the error below when testing the `rule-event-search` command though.

```
(.venv) hostname:detection-rules username$ python -m detection_rules dev test rule-event-search --cloud-id "my-cloud-id" --rule-id "e48236ca-b67a-4b4e-840c-fdc7782bc0c3"
█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█
es_user: elastic
es_password:
CLI Error (NoneType): Missing required --cloud-id or --elasticsearch-url
```

File: detection_rules/eswrap.py [hunk 4]
```diff
@@ -6,37 +6,56 @@
...
 
 COLLECTION_DIR = get_path('collections')
+MATCH_ALL = {'bool': {'filter': [{'match_all': {}}]}}
 
 
-def get_es_client(es_user, es_password, elasticsearch_url=None, cloud_id=None, **kwargs):
-    """Get an auth-validated elsticsearch client."""
-    assert elasticsearch_url or cloud_id, \
-        'You must specify a host or cloud_id to authenticate to an elasticsearch instance'
+def get_authed_es_client(cloud_id=None, elasticsearch_url=None, es_user=None, es_password=None, ctx=None, **kwargs):
```
Comment (rw-access) [inline detection_rules/eswrap.py:28]: not a fan of "authed"
```suggestion
def get_elasticsearch_client(cloud_id=None, elasticsearch_url=None, es_user=None, es_password=None, ctx=None, **kwargs):
```

File: detection_rules/devtools.py (hunk 3 above)
Comment (brokensound77) [reply detection_rules/devtools.py:251]: yea, pretty clean, huh

File: detection_rules/eswrap.py (hunk 4 above)
Comment (brokensound77) [reply detection_rules/eswrap.py:28]: wfm - I will make the change for kibana func too

**Comment by brokensound77** [comment]
nice find @threat-punter - should be resolved in c3a57c14fef7bf35e00f1d4c538e7ef6009e0948

Now rather than trying to re-auth a client in a nested invocation, it will use the existing client in `kwargs`. This allows (puts the onus) on the developer to re-pass auth kwargs and pop the existing client if a _new_ client is desired.

File: detection_rules/eswrap.py [hunk 5]
```diff
@@ -127,63 +148,198 @@ def _get_current_time(self, agent_hostname, index_str):
...
+        for rule, search_args in eql_searches.items():
+            try:
+                result = self.client.eql.search(**search_args)
+                survey_results[rule.id] = parse_unique_field_results(rule.type, rule.unique_fields, result)
+            except (elasticsearch.NotFoundError, elasticsearch.RequestError) as e:
+                survey_results[rule.id] = {'error_retrieving_results': True, 'error': e.info['error']['reason']}
+
+        for rule, async_id in async_searches.items():
+            result = async_client.get(async_id)['response']
+            survey_results[rule.id] = parse_unique_field_results(rule.type, rule.unique_fields, result)
```
Comment (brokensound77) [inline detection_rules/eswrap.py:269]: _may_ need to wrap this in exception handling

**Comment by threat-punter** [comment]
> nice find @threat-punter - should be resolved in [c3a57c1](https://github.com/elastic/detection-rules/commit/c3a57c14fef7bf35e00f1d4c538e7ef6009e0948)

Looking good @brokensound77 🙌

**Review by threat-punter — APPROVED** [review]
Tested all 4 command successfully. This is awesome.

File: detection_rules/devtools.py [hunk 6]
```diff
@@ -201,3 +206,146 @@ def license_check(ctx):
...
+        rule = Rule(rule_file, load_dump(rule_file))
+    else:
+        client_error('Must specify a rule file or rule ID')
+
+    if rule.query and rule.contents.get('language'):
+        if verbose:
+            click.echo(f'Searching rule: {rule.name}')
+
+        rule_lang = rule.contents.get('language')
+        language = None if rule_lang == 'kuery' else True if rule_lang == 'eql' else "lucene"
```
Comment (rw-access) [inline detection_rules/devtools.py:276]: split over multiple lines vs the double ternary operator

Comment (brokensound77) [reply detection_rules/devtools.py:276]: `"lucene"` was also a bug

```suggestion
        if rule_lang == 'kuery':
            language = None
        elif rule_lang == 'eql':
            language = True
        else:
            language = False
```

File: detection_rules/devtools.py [hunk 7]
```diff
@@ -201,3 +206,151 @@ def license_check(ctx):
...
+            click.echo(f'Searching rule: {rule.name}')
+
+        rule_lang = rule.contents.get('language')
+        if rule_lang == 'kuery':
+            language = None
+        elif rule_lang == 'eql':
+            language = True
+        else:
+            language = False
+        ctx.invoke(event_search, query=rule.query, index=rule.contents.get('index', '*'), language=language,
```
Comment (rw-access) [inline detection_rules/devtools.py:282]:
```suggestion
        ctx.invoke(event_search, query=rule.query, index=rule.contents.get('index', ['*']), language=language,
```

File: detection_rules/misc.py [hunk 8]
```diff
@@ -246,3 +248,90 @@ def getdefault(name):
...
+            client_ops_keys[c_type] = list(ops)
+
+        if not client_ops_dict:
+            raise ValueError(f'Unknown client: {client_type} in {func.__name__}')
+
+        client_ops = list(client_ops_dict.values())
+
+        @wraps(func)
+        @add_params(*client_ops)
+        def _wrapped(*args, **kwargs):
```
Comment (rw-access) [inline detection_rules/misc.py:295]: Thank you for doing it this way -- you know how i feel about interaction within a click callback, but instead wrapping the target function makes a lot more sense. +1 from me

File: requirements.txt [hunk 9]
```diff
@@ -6,7 +6,7 @@ requests==2.22.0
 Click==7.0
 PyYAML~=5.3
 eql~=0.9.5
-elasticsearch~=7.5.1
+elasticsearch~=7.9.1
```
Comment (rw-access) [inline requirements.txt:9]: loosen up the minor version
```suggestion
elasticsearch~=7.9
```

**Review by rw-access — APPROVED** [review]
two small nits, overall good

**Pull request merged**